File: Sem2/PO_project/Java/MapWindow.java
package Java;

import javax.swing.*;
import javax.swing.border.TitledBorder;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.util.ArrayList;

public class MapWindow extends JFrame implements KeyListener {
    public final int ROWS;
    public final int COLS;

    private final Game game;
    private final JButton[][] buttons;

    private final JPanel gameBoard;
    private final JPanel menu;
    private JTextArea logArea;

    private JComponent boardView;
    private HexBoardPanel hexPanel;

    public MapWindow(int rows, int cols) {
        this.ROWS = rows;
        this.COLS = cols;

        this.game = new Game(ROWS, COLS);
        this.buttons = new JButton[ROWS][COLS];

        this.gameBoard = new JPanel(new BorderLayout());
        this.menu = new JPanel(new BorderLayout());

        configureWindow();
        configurePanels();
        buildLayout();
        buildBoardView();
        game.initWorld();
        fillMenu();
        game.beginTurnLogs();
        showLogs(game.getTurnLogs());
        refreshBoard();
        
        this.addKeyListener(this);
        this.setFocusable(true);
    }

    @Override
    public void keyPressed(KeyEvent e) {
        int keyCode = e.getKeyCode();
        boolean processed = false;

        game.beginTurnLogs();

        int dir = -1;
        boolean isHex = game.getBoard().isHexagonal();

        if(!isHex){
            if (keyCode == KeyEvent.VK_UP) dir = 1;
            if (keyCode == KeyEvent.VK_DOWN) dir = 6;
            if (keyCode == KeyEvent.VK_LEFT) dir = 3;
            if (keyCode == KeyEvent.VK_RIGHT) dir = 4;
        } else {
        if (keyCode == KeyEvent.VK_W) dir = 0; // NW
        if (keyCode == KeyEvent.VK_E) dir = 1; // NE
        if (keyCode == KeyEvent.VK_A) dir = 2; // W
        if (keyCode == KeyEvent.VK_D) dir = 3; // E
        if (keyCode == KeyEvent.VK_Z) dir = 4; // SW
        if (keyCode == KeyEvent.VK_X) dir = 5; // SE
        }

        if (dir >= 0) {
            game.setHumanDirection(dir);
            processed = true;
        }

        switch (keyCode) {
            case KeyEvent.VK_P:
                game.activateHumanSpecial();
                processed = true;
                break;
            case KeyEvent.VK_S:
                game.logEvent(Config.save(game));
                showLogs(game.getTurnLogs());
                return;
            case KeyEvent.VK_R:
                game.logEvent(Config.load(game));
                refreshBoard();
                showLogs(game.getTurnLogs());
                return;
            case KeyEvent.VK_B:
                game.switchBoard();
                buildBoardView();
                refreshBoard();
                showLogs(game.getTurnLogs());
                return;
        }

        if (processed) {
            game.playTurn();
            refreshBoard();
            showLogs(game.getTurnLogs());
        }
    }

    @Override public void keyTyped(KeyEvent e) {}
    @Override public void keyReleased(KeyEvent e) {}

    private void configureWindow() {
        setTitle(("Game Map - Jakub Wintoch 208231"));
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLayout(new BorderLayout());
        setSize(1000,800);
        setLocationRelativeTo(null);      
    }

    private void configurePanels(){
        gameBoard.setBackground(Color.black);
        menu.setBackground(new Color(30,30,30));

        menu.setPreferredSize(new Dimension(1000, 180));

        TitledBorder boardBorder = BorderFactory.createTitledBorder("World");
        boardBorder.setTitleColor(Color.WHITE);
        gameBoard.setBorder(boardBorder);

        TitledBorder menuBorder = BorderFactory.createTitledBorder("Logs and Controls");
        menuBorder.setTitleColor(Color.WHITE);
        menu.setBorder(menuBorder);
    }

    private void buildLayout(){
        add(gameBoard, BorderLayout.CENTER);
        add(menu, BorderLayout.SOUTH);
    }

    private void onCellClick(int r, int c) {
            if (game.getOrganismAt(r, c) != null) {
        game.logEvent("This cell is occupied. Cannot add new organism.");
        showLogs(game.getTurnLogs());
        return;
    }

    String[] options = {
        "Wolf", "Sheep", "Fox", "Turtle", "Antelope",
        "Grass", "Guarana", "Milkweed", "NightShade"
    };

    String choice = (String) JOptionPane.showInputDialog(
        this,
        "Choose organism:",
        "Add organism",
        JOptionPane.PLAIN_MESSAGE,
        null,
        options,
        options[0]
    );

    if (choice == null) return;

    Organism org = game.createOrganismByName(choice, r, c);
        if (org != null) {
            game.addOrganismAt(r, c, org);
            game.logEvent("Dodano " + choice + " na [" + r + "," + c + "]");
            refreshBoard();
            showLogs(game.getTurnLogs());
        }
    }

    // private void fillBoard() {
    //     for (int r = 0; r < ROWS; r++) {
    //         for (int c = 0; c < COLS; c++) {
    //             JButton button = new JButton();
    //             button.setBackground(Color.black);
    //             button.setOpaque(true);
    //             button.setBorder(BorderFactory.createLineBorder(Color.BLACK));
    //             button.setFocusable(false);

    //             final int rr = r, cc = c;
    //             button.addActionListener(e -> onCellClick(rr, cc));

    //             buttons[r][c] = button;
    //             gameBoard.add(button);
    //         }
    //     }
    // }

    public void refreshBoard() {
        if (game.getBoard().isHexagonal()) {
            if (hexPanel != null) {
                hexPanel.repaint();
            }
        } else {
            for (int r = 0; r < ROWS; r++) {
                for (int c = 0; c < COLS; c++) {
                    Organism org = game.getOrganismAt(r, c);
                    if (org != null) {
                        buttons[r][c].setBackground(org.getColor());
                    } else {
                        buttons[r][c].setBackground(Color.black);
                    }
                }
            }
        }
        repaint();
    }

    private void buildBoardView() {
        if (boardView != null) {
            gameBoard.remove(boardView);
        }

        if (game.getBoard().isHexagonal()) {
            hexPanel = new HexBoardPanel(game, 16);
            hexPanel.setOnCellClick(this::onCellClick);
            boardView = new JScrollPane(hexPanel);
        } else {
            JPanel grid = new JPanel(new GridLayout(ROWS, COLS));
            for (int r = 0; r < ROWS; r++) {
                for (int c = 0; c < COLS; c++) {
                    JButton button = new JButton();
                    button.setBackground(Color.black);
                    button.setOpaque(true);
                    button.setBorder(BorderFactory.createLineBorder(Color.BLACK));
                    button.setFocusable(false);

                    final int rr = r, cc = c;
                    button.addActionListener(e -> onCellClick(rr, cc));

                    buttons[r][c] = button;
                    grid.add(button);
                }
            }
            boardView = grid;
        }

        gameBoard.add(boardView);
        gameBoard.revalidate();
        gameBoard.repaint();
    }

    private void fillMenu() {
        menu.setLayout(new BorderLayout());

        JPanel legendPanel = new JPanel(new GridLayout(0, 2, 10, 2));
        legendPanel.setOpaque(false);
        legendPanel.setPreferredSize(new Dimension(400, 160));

        Object[][] legendInfo = {
            {"Human", new Color(0, 0, 255)},
            {"Wolf", new Color(128, 128, 128)},
            {"Sheep", new Color(255, 255, 255)},
            {"Fox", new Color(255, 165, 0)},
            {"Turtle", new Color(0, 128, 0)},
            {"Antelope", new Color(160, 82, 45)},
            {"Grass", new Color(0, 255, 0)},
            {"Guarana", new Color(255, 105, 180)},
            {"Milkweed", new Color(255, 255, 0)},
            {"Nightshade", new Color(128, 0, 128)},
            {"Hogweed", new Color(0, 255, 255)}
        };

        for (Object[] item : legendInfo){
            String name = (String) item[0];
            Color color = (Color) item[1];

            JPanel itemPanel = new JPanel(new FlowLayout(FlowLayout.LEFT, 10, 0));
            itemPanel.setOpaque(false);

            JPanel colorBox = new JPanel();
            colorBox.setBackground(color);
            colorBox.setPreferredSize(new Dimension(10, 10));
            colorBox.setBorder(BorderFactory.createLineBorder(Color.BLACK));

            JLabel label = new JLabel(name);
            label.setForeground(Color.WHITE);

            itemPanel.add(colorBox);
            itemPanel.add(label);
            legendPanel.add(itemPanel);
        }
        
        menu.add(legendPanel, BorderLayout.WEST);

        logArea = new JTextArea();
        logArea.setEditable(false);
        logArea.setBackground(new Color(20, 20, 20));
        logArea.setForeground(Color.WHITE);
        logArea.setLineWrap(true);
        logArea.setWrapStyleWord(true);
        logArea.setFocusable(false);
        
        JPanel logPanel = new JPanel(new BorderLayout());
        logPanel.add(logArea, BorderLayout.CENTER);

        menu.add(logPanel, BorderLayout.CENTER);
    }

    private void showLogs(ArrayList<String> logs) {
        clearLogs();

        String specialState = game.isHumanSpecialActive() ? "ON" : "OFF";
        String canUse = game.canUseAbility() ? "YES" : "NO";
        int cooldown = game.getHumanCooldown();

        addLog("Ability: " + specialState
                + " | Can use: " + canUse
                + " | Cooldown: " + cooldown
                + " | Strength Human: " + game.getHumanStrength());

        addLog("Save to file: S | Load from file: R");
        addLog("To switch board press: B");

        for (String line : logs) {
            addLog(line);
        }
    }

    public void addLog(String message) {
        logArea.append(message + "\n");
        logArea.setCaretPosition(logArea.getDocument().getLength());
    }

    public void clearLogs() {
        logArea.setText("");
    }
}

File: Sem2/PO_project/python/map_window.py
from __future__ import annotations

import tkinter as tk
from tkinter import messagebox, simpledialog

from python.config import load as load_game
from python.config import save as save_game
from python.game import Game
from python.hex_board_panel import HexBoardPanel


class MapWindow(tk.Tk):
    def __init__(self, rows: int, cols: int) -> None:
        super().__init__()
        self.ROWS = rows
        self.COLS = cols
        self.game = Game(rows, cols)
        self.board_widget = None
        self.square_canvas = None
        self.cell_size = 32

        self._configure_window()
        self._configure_panels()
        self._build_layout()
        self._build_board_view()
        self.game.init_world()
        self._fill_menu()
        self.game.begin_turn_logs()
        self._show_logs(self.game.get_turn_logs())
        self.refresh_board()

        self.bind_all("<Key>", self._on_key_press)
        self.focus_force()

    def _configure_window(self) -> None:
        self.title("Game Map - Jakub Wintoch 208231")
        self.geometry("1000x800")
        self.configure(background="#111111")

    def _configure_panels(self) -> None:
        self.game_board = tk.Frame(self, background="black", highlightbackground="#444444", highlightthickness=1)
        self.menu = tk.Frame(self, background="#1e1e1e", height=180, highlightbackground="#444444", highlightthickness=1)

    def _build_layout(self) -> None:
        self.game_board.pack(fill="both", expand=True)
        self.menu.pack(fill="x", side="bottom")

    def _build_board_view(self) -> None:
        if self.board_widget is not None:
            self.board_widget.destroy()

        if self.game.get_board().is_hexagonal():
            self.board_widget = HexBoardPanel(self.game_board, self.game, size=16)
            self.board_widget.set_on_cell_click(self._on_cell_click)
            self.board_widget.pack(fill="both", expand=True)
        else:
            self.square_canvas = tk.Canvas(self.game_board, background="black", highlightthickness=0)
            self.square_canvas.pack(fill="both", expand=True)
            self.square_canvas.bind("<Button-1>", self._on_square_click)
            self.board_widget = self.square_canvas

        self.game_board.update_idletasks()

    def _color_to_hex(self, color: tuple[int, int, int]) -> str:
        return "#%02x%02x%02x" % color

    def _on_square_click(self, event) -> None:
        row = event.y // self.cell_size
        col = event.x // self.cell_size
        if self.game.is_inside(row, col):
            self._on_cell_click(row, col)

    def _on_cell_click(self, row: int, col: int) -> None:
        if self.game.get_organism_at(row, col) is not None:
            self.game.log_event("This cell is occupied. Cannot add new organism.")
            self._show_logs(self.game.get_turn_logs())
            return

        options = ["Wolf", "Sheep", "Fox", "Turtle", "Antelope", "Grass", "Guarana", "Milkweed", "NightShade"]
        choice = simpledialog.askstring("Add organism", "Choose organism:\n" + ", ".join(options), parent=self)
        if choice is None:
            return

        normalized = choice.strip().lower()
        mapping = {item.lower(): item for item in options}
        choice = mapping.get(normalized)
        if choice is None:
            messagebox.showerror("Invalid choice", "Unknown organism name.")
            return

        org = self.game.create_organism_by_name(choice, row, col)
        if org is not None:
            self.game.add_organism_at(row, col, org)
            self.game.log_event(f"Dodano {choice} na [{row},{col}]")
            self.refresh_board()
            self._show_logs(self.game.get_turn_logs())

    def refresh_board(self) -> None:
        if self.game.get_board().is_hexagonal():
            if isinstance(self.board_widget, HexBoardPanel):
                self.board_widget.redraw()
        else:
            if self.square_canvas is None:
                return
            self.square_canvas.delete("all")
            self.square_canvas.configure(width=self.COLS * self.cell_size, height=self.ROWS * self.cell_size)
            for r in range(self.ROWS):
                for c in range(self.COLS):
                    x1 = c * self.cell_size
                    y1 = r * self.cell_size
                    x2 = x1 + self.cell_size
                    y2 = y1 + self.cell_size
                    org = self.game.get_organism_at(r, c)
                    fill = self._color_to_hex(org.get_color()) if org is not None else "#000000"
                    self.square_canvas.create_rectangle(x1, y1, x2, y2, fill=fill, outline="#282828")

    def _fill_menu(self) -> None:
        for child in self.menu.winfo_children():
            child.destroy()

        legend_panel = tk.Frame(self.menu, background="#1e1e1e")
        legend_panel.pack(side="left", padx=10, pady=10)

        legend_info = [
            ("Human", (0, 0, 255)),
            ("Wolf", (128, 128, 128)),
            ("Sheep", (255, 255, 255)),
            ("Fox", (255, 165, 0)),
            ("Turtle", (0, 128, 0)),
            ("Antelope", (160, 82, 45)),
            ("Grass", (0, 255, 0)),
            ("Guarana", (255, 105, 180)),
            ("Milkweed", (255, 255, 0)),
            ("Nightshade", (128, 0, 128)),
            ("Hogweed", (0, 255, 255)),
        ]

        for name, color in legend_info:
            row = tk.Frame(legend_panel, background="#1e1e1e")
            row.pack(anchor="w")
            box = tk.Frame(row, background=self._color_to_hex(color), width=10, height=10, highlightbackground="#000000", highlightthickness=1)
            box.pack(side="left", padx=(0, 8), pady=2)
            label = tk.Label(row, text=name, foreground="white", background="#1e1e1e")
            label.pack(side="left")

        self.log_area = tk.Text(self.menu, height=8, background="#141414", foreground="white", insertbackground="white", wrap="word", borderwidth=0)
        self.log_area.pack(side="left", fill="both", expand=True, padx=10, pady=10)
        self.log_area.configure(state="disabled")

    def _show_logs(self, logs: list[str]) -> None:
        self._clear_logs()

        special_state = "ON" if self.game.is_human_special_active() else "OFF"
        can_use = "YES" if self.game.can_use_ability() else "NO"
        cooldown = self.game.get_human_cooldown()

        self._add_log(f"Ability: {special_state} | Can use: {can_use} | Cooldown: {cooldown} | Strength Human: {self.game.get_human_strength()}")
        self._add_log("Save to file: S | Load from file: R")
        self._add_log("To switch board press: B")

        for line in logs:
            self._add_log(line)

    def _add_log(self, message: str) -> None:
        self.log_area.configure(state="normal")
        self.log_area.insert("end", message + "\n")
        self.log_area.see("end")
        self.log_area.configure(state="disabled")

    def _clear_logs(self) -> None:
        self.log_area.configure(state="normal")
        self.log_area.delete("1.0", "end")
        self.log_area.configure(state="disabled")

    def _on_key_press(self, event) -> None:
        key = event.keysym
        processed = False

        self.game.begin_turn_logs()

        direction = -1
        is_hex = self.game.get_board().is_hexagonal()

        if not is_hex:
            if key == "Up":
                direction = 1
            elif key == "Down":
                direction = 6
            elif key == "Left":
                direction = 3
            elif key == "Right":
                direction = 4
        else:
            mapping = {"w": 0, "e": 1, "a": 2, "d": 3, "z": 4, "x": 5}
            direction = mapping.get(key.lower(), -1)

        if direction >= 0:
            self.game.set_human_direction(direction)
            processed = True

        lower_key = key.lower()
        if lower_key == "p":
            self.game.activate_human_special()
            processed = True
        elif lower_key == "s":
            self.game.log_event(save_game(self.game))
            self._show_logs(self.game.get_turn_logs())
            return
        elif lower_key == "r":
            self.game.log_event(load_game(self.game))
            self.refresh_board()
            self._show_logs(self.game.get_turn_logs())
            return
        elif lower_key == "b":
            self.game.switch_board()
            self._build_board_view()
            self.refresh_board()
            self._show_logs(self.game.get_turn_logs())
            return

        if processed:
            self.game.play_turn()
            self.refresh_board()
            self._show_logs(self.game.get_turn_logs())
